server/src/test/java/keywhiz/service/resources/BatchSecretDeliveryResourceTest.java

```diff
@@ -0,0 +1,163 @@
+            null, null, null, null, 0, 1L, NOW, null);
+
+    final Secret secret2 = new Secret(2, "secret2_name", null, () -> "secret_value2", "checksum2", NOW, null, NOW, null,
+            null, null, null, 0, 1L, NOW, null);
+
+
+    @Before
+    public void setUp() {
+        batchSecretDeliveryResource = new BatchSecretDeliveryResource(secretController, aclDAO, clientDAO);
+    }
+
+    @Test
+    public void returnsSingleSecretWhenAllowed() throws Exception {
+        SanitizedSecret sanitizedSecret = SanitizedSecret.fromSecret(secret);
+
+        ImmutableList<String> secretnames = ImmutableList.of(sanitizedSecret.name());
+        BatchSecretRequest req = BatchSecretRequest.create(secretnames);
+
+        when(aclDAO.getBatchSanitizedSecretsFor(client, secretnames))
+                .thenReturn(List.of(sanitizedSecret));
+        when(secretController.getSecretsByName(secretnames))
+                .thenReturn(List.of(secret));
+
+        List<SecretDeliveryResponse> response = batchSecretDeliveryResource.getBatchSecret(client, req);
+        assertThat(response).isEqualTo(ImmutableList.of(SecretDeliveryResponse.fromSecret(secret)));
+    }
+
+    @Test
+    public void returnsMultipleSecretsWhenAllowed() throws Exception {
+        SanitizedSecret sanitizedSecret = SanitizedSecret.fromSecret(secret);
+        SanitizedSecret sanitizedSecret2 = SanitizedSecret.fromSecret(secret2);
+
+        ImmutableList<String> secretnames = ImmutableList.of(sanitizedSecret.name(), sanitizedSecret2.name());
+        BatchSecretRequest req = BatchSecretRequest.create(secretnames);
+
+        when(aclDAO.getBatchSanitizedSecretsFor(client, secretnames))
+                .thenReturn(List.of(sanitizedSecret, sanitizedSecret2));
+        when(secretController.getSecretsByName(secretnames))
+                .thenReturn(List.of(secret, secret2));
+
+        List<SecretDeliveryResponse> response = batchSecretDeliveryResource.getBatchSecret(client, req);
+
+        // The order doesn't matter
+        boolean matches = false;
+
+        if (response.equals(ImmutableList.of(SecretDeliveryResponse.fromSecret(secret), SecretDeliveryResponse.fromSecret(secret2)))) {
+            matches = true;
+        } else if (response.equals(ImmutableList.of(SecretDeliveryResponse.fromSecret(secret2), SecretDeliveryResponse.fromSecret(secret)))) {
+            matches = true;
+        }
+
+        assertThat(matches).isTrue();
+    }
+
+    @Test(expected = NotFoundException.class)
+    public void returnsNotFoundWhenClientDoesNotExist() throws Exception {
+        when(aclDAO.getBatchSanitizedSecretsFor(client, ImmutableList.of(secret.getName()))).thenReturn(ImmutableList.<SanitizedSecret>of());
+        when(clientDAO.getClientByName(client.getName())).thenReturn(Optional.empty());
+        when(secretController.getSecretsByName(ImmutableList.of(secret.getName())))
```

Comment:
This test should be set up with an existing secret, but missing client--currently both the client and the secret are missing, so it's hard to tell which failure case is actually being triggered.